Enable image builds with badge and add os-ref param for manual builds

dbast opened the pull request:
This updates the existing build action to:
* add an os-ref input parameter to be able to do manual workflow_dispatch builds where any seedsigner branch can be combined with any seedsigner-os ref (tag, branch, sha1). (default = the default branch of both repos)
* produce unique (git describe based)image file names that have the seedsigner and seedsigner-os version encoded, e.g. seedsigner_os.os0.6.0-61-g9fafebe_sw0.7.0-40-g4b7c895.pi0.img to know exactly what version an image is based on.
* simplify the cache action step: restore + save is automatically done by the plain action and its included post-action step. Some testing shows that the cache action reduces the build time to 50% (~30mins).
* ~~enable daily builds = daily PoW that everything works, sources exist, and builds fine together. And it also enables easily testing of merged PRs of the last days for everybody without locally building anything or waiting for a manual workflow_dispatch run. (this also keeps the cache warm, so succeeding triggered builds are a lot faster).~~ one build per merge
* ~~lower the retention-days to 7 (default 90 days), to not consume to much artifact space with the daily builds enabled.~~ newer builds overwrite older builds in case the artifact cache is full

For more details see inline comments.

Succeeding PRs can also enable builds for every merge to the dev/main branch or for triggering builds when a tag is added.

some links: 
* A full run of the build workflow triggered via manual workflow_dispatch from the branch of the PR can be seen here https://github.com/dbast/seedsigner/actions/runs/6259494273 
* this one p0 run finished in 22 mins due to caching https://github.com/dbast/seedsigner/actions/runs/6259494273/job/16995636719 
* this is a cron triggered run with disabled fork protection, reduced build matrix and caching https://github.com/dbast/seedsigner/actions/runs/6218622125/job/16875333434

## Description

_Describe the change simply. Provide a reason for the change._

_Include screenshots of any new or modified screens (or at least explain why they were omitted)_

This pull request is categorized as a:

- [ ] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Documentation
- [x] Other

## Checklist

- [ ] I’ve run `pytest` and made sure all unit tests pass before sumbitting the PR

If you modified or added functionality/workflow, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [x] N/A

I have tested this PR on the following platforms/os:

- [ ] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [ ] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board
- [ ] Other


Note: Keep your changes limited in scope; if you uncover other issues or improvements along the way, ideally submit those as a separate PR. The more complicated the PR the harder to review, test, and merge.

kdmukai commented:
Out of my area of expertise, but this sounds awesome. Opens the door for having less technical people try out PRs or the latest merged into `dev`, ya?

A few thoughts @dbast:
* Is the daily build necessary? We have very infrequent merges so it makes more sense to me to just have:
  * A build for each PR
  * A build for each merge
* Expiration: I think it would be more useful to have the resulting images stick around for longer.
  * For PRs: Updated/replaced on PR updates, remains available until the PR is merged or closed (is that possible?)
  * For merges: Keep the last n merges?

dbast commented:
@kdmukai Rebased the PR and addressed your review findings: one build per merged PR, PR image builds only on demand as resource hungry (can be triggered via comment change, see code), retention days increased to maximum possible.

newtonick commented:
ACK
